### PR TITLE
Dev 1737 refactor nextflow role

### PR DESCRIPTION
**Pull request by GitEdvard**
Purpose:
The seqreports role is referred to from sysman, when provisioning the seqreports-dev machine. However, it's not possible to run Seqreports-dev as long as nextflow is dependent on ngi_pipeline. Therefore, the dependencies between nextflow and ngi_pipeline roles are reversed. 

Details:
A portion of the tasks within the nextflow role is related to update e.g. sourceme_common file, which in turn is created by the ngi_pipeline role. These tasks (that is configuring environment variables)  are moved to the ngi_pipeline role and gathered in the file "config_nextflow.yml". 

Validations:
The nextflow role had dependencies on ngi_pipeline and nf-core roles. I have run ngi_pipeline role before and after this refactoring, without errors. Regarding the nf-core role, I cannot go past the "nf-core download rnaseq v3.0" task, neither before or after the refactoring.

### CODE REVIEW

**Review by b97pla — CHANGES_REQUESTED** [review]
Great work! I think this largely looks good. 

However, I would object to the part where `ngi_pipeline` role adds the `nextflow` role's parameters to the source files etc. I think that is the wrong place for that. Basically, as it is now, this is required because the `ngi_pipeline` role is probably not the proper place for these files to be initially set up in the first place.

I would suggest, either creating a separate role that will set up the common files needed for several roles (`sourceme_...`, `deployed_tools...`, `create_directory...` etc.) or to just move these tasks from `ngi_pipeline` to `anaconda` so that they are in place when the `nextflow`  role is run. 

What do you think?

**Comment by GitEdvard** [comment]
Just a control question, could it ever be meaningful to run the new environment/configuration role together with another role which is not the ngi-pipeline role? Generally I think it's a good approach to wait to do such refactoring until we have a specific need for it, in which time we probably have more detailed knowledge of the requirements. I have nothing against to do this refactoring now though (you might sit on knowledge that I don't have). 

Moving these tasks to anaconda shouldn't be feasible for the sake of seqreports role, which is dependent on anaconda as well.

**Comment by b97pla** [comment]
I would say so. All roles depend on `deployed_tools` and several roles use the other files as well, so it's not entirely logical that they are set up in the `ngi_pipeline` role to begin with and it could make sense to have a dedicated role set these up.

My thinking with `anaconda` was precisely that all roles depend on it so it will always be run first and therefore it would be a shortcut to set up the files in that role. But I may be misunderstanding the premise for the `seqreports` role on `sysman`. Also, I'm more in favor of my first suggestion..

**Comment by GitEdvard** [comment]
Ok, thanks for the clarification. I refactor the configuration into a separate role.

**Comment by GitEdvard** [comment]
The refactoring is done :heavy_check_mark:  

* configure_base_sw runs without ngi_pipeline role. 
* ngi_pipeline is dependent on configure_base_sw (in meta dir), and runs without errors. 

Do you think the name configure_base_sw is relevant? Please check if all things is OK.

**Comment by GitEdvard** [comment]
Please not that this PR goes to the staging branch 1737-seqreports-branch. I don't want to crash the current validation. Otherwise, I think this PR should go to the main branch (monthly?) rather quickly.

File: roles/configure_base_sw/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    dest: "{{ ngi_pipeline_conf }}/{{ bash_env_common_script }}"
+    line: "export PATH={{ anaconda_path }}/bin:$PATH"
+    backup: no
+
+- name: Deploy bash environment variables for {{ site }}
+  template:
+    src: "{{ bash_env_site_script }}.j2"
+    dest: "{{ ngi_pipeline_conf }}/{{ bash_env_script }}"
+
+- name: Add py3 anaconda env path to sourceme script for {{ site }}
```
Comment (b97pla) [inline roles/configure_base_sw/tasks/main.yml:28]: move to `ngi_pipeline`

File: roles/configure_base_sw/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  lineinfile:
+    dest: "{{ ngi_pipeline_conf }}/{{ bash_env_common_script }}"
+    line: "export PATH={{ anaconda_path }}/envs/{{ NGI_venv_py3_name }}/bin:$PATH"
+    backup: no
+
+- name: Try getting anaconda version
```
Comment (b97pla) [inline roles/configure_base_sw/tasks/main.yml:34]: move to `anaconda`

File: roles/configure_base_sw/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  shell: "conda --version"
+  register: conda_version
+  ignore_errors: true
+
+- name: Store anaconda version in deployment
```
Comment (b97pla) [inline roles/configure_base_sw/tasks/main.yml:39]: move to `anaconda`

File: install.yml [hunk 4]
```diff
@@ -70,7 +70,6 @@
   roles:
     - { role: ngi_pipeline, tags: ngi_pipeline }
     - { role: func_accounts, tags: func_accounts }
-    - { role: nextflow, tags: nextflow }
```
Comment (b97pla) [inline install.yml:73]: this role should not be removed

File: roles/configure_base_sw/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  shell: "ls {{ nextflow_dest }}/workfiles/framework/ | tail -1"
+  register: nextflow_ver
+
+- name: Add NextFlow to $PATH
+  lineinfile:
+    dest: "{{ ngi_pipeline_conf }}/{{ bash_env_common_script }}"
+    line: "export PATH={{ root_path }}/sw/nextflow:$PATH"
+    backup: no
+
+- name: Create miarka config
```
[the remark this answers is not in the document]
Comment (b97pla) [reply roles/configure_base_sw/tasks/main.yml:55]: the template should remain in the Nextflow role

File: roles/nextflow/meta/main.yml [hunk 6]
```diff
@@ -1,2 +0,0 @@
-dependencies:
-  - { role: ngi_pipeline, tags: ngi_pipeline, environment: "{{ tools_path }}" }
```
Comment (b97pla) [inline roles/nextflow/meta/main.yml:2]: it should depend on `configure_base_sw`

File: roles/nf-core/meta/main.yml [hunk 7]
```diff
@@ -1,2 +1,2 @@
 dependencies:
-  - { role: nextflow, tags: nextflow }
+  - { role: ngi_pipeline, tags: ngi_pipeline, environment: "{{ tools_path }}" }
```
Comment (b97pla) [inline roles/nf-core/meta/main.yml:2]: I believe it should be depending on `Nextflow` and `anaconda`

File: roles/ngi_pipeline/meta/main.yml [hunk 8]
```diff
@@ -1,2 +1,2 @@
 dependencies:
-  - { role: anaconda, tags: anaconda }
+  - { role: configure_base_sw, tags: configure_base_sw }
```
Comment (b97pla) [inline roles/ngi_pipeline/meta/main.yml:2]: also `anaconda`

File: roles/configure_base_sw/README.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
+configure_base_sw
+=========
+
+Configure the base software (currently nextflow and anaconda), and setup environment variables for these. Creates file for deployed_tool_versions, which is used by most other roles. Deploys sourceme_common.sh and sourceme_site.sh.
```
Comment (b97pla) [inline roles/configure_base_sw/README.md:4]:
```suggestion
Creates file for deployed_tool_versions, which is used by most other roles. Deploys sourceme_common.sh and sourceme_site.sh.
```

**Comment by GitEdvard** [comment]
Now, let's give it another try :)

File: roles/ngi_pipeline/tasks/main.yml [hunk 10]
```diff
@@ -1,5 +1,11 @@
 ---
 
+- name: Add py3 anaconda env path to sourceme script for {{ site }}
```
Comment (b97pla) [inline roles/ngi_pipeline/tasks/main.yml:3]: personally, I would put this task after the environment has been created but it's not that important

**Review by b97pla — APPROVED** [review]
:+1: great work, I'm happy 😃

**Pull request merged**